### DISCUSSION TITLE
NPM workspaces

### DISCUSSION THREAD

**Discussion by guimochila** [eslint/eslint#16817] [Q&A] [answered]
I have been searching about how to setup eslint in a NPM workspaces but I get many different answers, so I would like to ask here if I have eslint installed in the root package.json, should I need to install eslint on each package in the workspace?

Thank you in advance.

**Comment by kecrily** [comment, marked as answer]
There is no need to install eslint in each package, just install eslint in the root directory and configure the appropriate glob pattern.

**Reply by guimochila** [reply]
Thanks @kecrily

**Comment by bmish** [comment]
I don't know if there is or should be an official recommendation. There are multiple valid approaches.

My preference (and an approach I have used/seen commonly working in large monorepos) is to install ESLint inside each package of the workspace that needs it. Essentially, any package with JavaScript files installs and runs its own linting, the same way it does with its own tests.

Benefits of this:
* Each package can be responsible for and customize its own dependencies, scripts, and configuration as needed, staying isolated without affecting the rest of the workspace. Note that there are still ways to enforce consistency across the workspace like with [Yarn Constraints](https://yarnpkg.com/features/constraints), [check-dependency-version-consistency](https://github.com/bmish/check-dependency-version-consistency) (tool I built for this), [npm-package-json-lint](https://github.com/tclindner/npm-package-json-lint).
* Each package is more portable. That is, if a package needed to be moved out of the workspace, it would be easy to do so without a lot of untangling.
* A package should declare all of its own dependencies. This enables better dependency analysis and other static analysis tools like [node/no-extraneous-require](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-extraneous-require.md) which attempts to detect missing dependencies, and so I don't consider relying on the workspace to provide dependencies to be best practice. 
* ESLint is just one kind of linter. There can be other linters, testers, or requirements present, varying based on the package. Each package can be responsible for enforcing its requirements, instead of having to make the top-level workspace root package.json aware of all the potential requirements of every package.
* CI can just run the `lint` / `test` script(s) from each package in separate build parts in parallel, instead of running everything together.

**Reply by guimochila** [reply]
Thank you @bmish! I like this approach too, so in this case there will be no reason to have eslint installed in the root package? I will do some tests with tools like lint-staged, husky. This always has been a tricky as you mentioned there is no official documentation about it.

**Reply by bmish** [reply]
With my approach, I would add ESLint to the top-level workspace root package.json only if there are JS files in that location to lint, since again, each package is in charge of linting its own files. If there are no JS files in the workspace root, then no need for ESLint at that level. I do however typically run other types of workspace-wide checks in the workspace root, for example:

```json
{
    "lint": "npm-run-all lint:*",
    "lint:dependency-versions": "check-dependency-version-consistency",
    "lint:package-json": "npmPkgJsonLint ."
}
```

**Reply by guimochila** [reply]
Cool! It helped a lot to get this "right", at least to get to know more about it when comes to monorepos. I also was thinking about another approach where you create your own lint package with your own configuration inside the monorepo and it is used by other packages so you make sure the eslint version is consistent across all workspaces.

**Reply by bmish** [reply]
Yeah! I've also used that pattern of having a private (unpublished) `eslint-plugin-[my-monorepo-name]` which exports a shared linting config that every other package in the workspace can extend from. This can also make it easy to create new lint rules that are specific to your workspace (although, I always try to contribute new lint rules to open source ESLint plugins when they are sufficiently-generic).

**Reply by kecrily** [reply]
I'm used to using my own generic configuration, so I just install ESLint and shared config in the root directory. It all depends on your specific needs

**Reply by guimochila** [reply]
That was a good insight as there is no official recommendation. I was just wondering a good way of keeping consistency across multiple packages and this discussion has helped a lot. Thanks @kecrily and @bmish
